File: Travel managment system - C#/projekt/BiuroPodrozy.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml.Serialization;

namespace projekt
{
    /// <summary>
    /// Klasa BiuroPodróży przechowująca podróże, korzystająca z interfejsów IBiuro oraz ICloneable
    /// </summary>
    [Serializable]
    [XmlInclude(typeof(Podroze))]
    
    public class BiuroPodrozy: IBiuro,ICloneable
    {
        /// <summary>
        /// Identyfikator (klucz) o nazwe BiuroID
        /// </summary>
        [Key]
        public int BiuroId { get; set; }

        /// <summary>
        /// Zmodyfikowane pole loty, które jest listą
        /// </summary>
        public virtual List<Podroze> loty { get; set; }
        

        /// <summary>
        /// Nazwa biura podróży
        /// </summary>
        private string nazwa;
        /// <summary>
        /// Hermetyzacja pola nazwa
        /// </summary>
        public string Nazwa { get => nazwa; set => nazwa = value; }
        /// <summary>
        /// Konstruktor nieparametryczny tworzący nową listę 'loty'
        /// </summary>
        public BiuroPodrozy() {
            loty = new List<Podroze>();
        }

        /// <summary>
        /// Konstruktor parametryczny ustawiający pole 'nazwa'
        /// </summary>
        /// <param name="nazwa">Nazwa biura podróży</param>
        public BiuroPodrozy(string nazwa) : this()
        {
            this.Nazwa = nazwa; 
        }

        /// <summary>
        /// Metoda zapisująca do bazy danych
        /// </summary>
        public void ZapiszDoBazy()
        {
            using (Model1 db = new Model1())
            {
                db.BiuroPod.Add(this);
                db.SaveChanges();
            }
        }
        /// <summary>
        /// Metoda odczytująca baze danych
        /// </summary>
        public static BiuroPodrozy OdczytZBazy()
        {
            using (Model1 db = new Model1())
            {
                int maxindex = db.BiuroPod.Max(zz => zz.BiuroId);
                var zbaza = db.BiuroPod.Find(maxindex); // Where(r => r.zespolId == maxindex).First();
                BiuroPodrozy z = new BiuroPodrozy();
                z.Nazwa = zbaza.Nazwa;
                z.loty = zbaza.loty;
                return z;
            }
        }

        /// <summary>
        /// Metoda dodająca wybraną podróż
        /// </summary>
        /// <param name="t">Wybrany lot do dodania</param>
        public void DodajLot(Podroze t)
        {
            loty.Add(t);
        }

        /// <summary>
        /// Metoda usuwająca wybraną podróż
        /// </summary>
        /// <param name="t">Wybrany lot do usunięcia</param>
        //Usuwanie lotu
        public void UsunLot(Podroze t)
        {
            loty.Remove(t);
        }

        /// <summary>
        /// Metoda usuwająca podróż z danym PESELEM
        /// </summary>
        /// <param name="PESEL">Wybrany PESEL</param>
        public void UsunLot(string PESEL)
        {
            foreach (Podroze l in loty)
            {
                if (l.PESEL == PESEL)
                {
                    loty.Remove(l);
                    break;
                }
            }
        }

        /// <summary>
        /// Metoda sprawdzajaca czy dana podróż jest już w biurze podróży
        /// </summary>
        /// <param name="podroz">wybrana podróż</param>
        /// <returns>true jeśli podroz jest już dodana albo false jeśli nie jest </returns>
        public bool CzyPodrozJestWBiurzePodrozy(Podroze podroz)
        {
            foreach (Podroze pod in loty)
                if (pod.Equals(podroz))
                    return true;
            return false;
        }

        /// <summary>
        /// Metoda, która wyszukuje podróż przez podanie nazwy lotniska z którego jest wylot
        /// </summary>
        /// <param name="wylot">Nazwa lotniska</param>
        /// <returns>Zwraca listę podróży</returns>
        public List<Podroze> WyszukajPodroze(string wylot)
        {
            List<Podroze> newList = new List<Podroze>();
            newList = loty.FindAll(c => c.Wylot.ToString().Equals(wylot));

            return newList;
        }

        /// <summary>
        /// Metoda, która sortuje loty (Po sygnaturze)
        /// </summary>
        public void Sortuj()
        {
            loty.Sort();
        }

        /// <summary>
        /// Metoda, która sortuje loty według peselu
        /// </summary>
        public void SortujPoPESEL()
        {
            loty.Sort(new PESELComparator());
        }

        /// <summary>
        /// Metoda zapisująca BiuroPodrozy do pliku xml 
        /// </summary>
        /// <param name="nazwa">Nazwa pliku XML</param>
        /// <param name="z">Biuro podróży do zapisania</param>
        public static void ZapiszXML(string nazwa, BiuroPodrozy z)
        {
            var serializer = new XmlSerializer(typeof(BiuroPodrozy));
            var fs = new StreamWriter(nazwa);
            serializer.Serialize(fs, z);
            fs.Close();
        }

        /// <summary>
        /// Metoda odczytująca BiuroPodrozy z pliku xml
        /// </summary>
        /// <param name="nazwa">Nazwa pliku xml do odczytania</param>
        /// <returns>Zwraca odczytane biuro podróży</returns>
        public static BiuroPodrozy OdczytajXML(string nazwa)
        {
            BiuroPodrozy noweBiuro;
            var xf = new XmlSerializer(typeof(BiuroPodrozy));
            var fs = new StreamReader(nazwa);
            noweBiuro = (BiuroPodrozy)xf.Deserialize(fs);
            fs.Close();
            return noweBiuro;
        }

        /// <summary>
        /// Metoda tworząca płytką kopię obiektu BiuroPodrozy
        /// </summary>
        /// <returns>Płytka kopia obiektu</returns>
        public object Clone()
        {
            return this.MemberwiseClone();
        }

        /// <summary>
        /// Metoda tworząca głęboką kopię obiektu BiuroPodrozy
        /// </summary>
        /// <returns>Głęboka kopia obiektu</returns>
        public BiuroPodrozy DeepCopy()
        {
            BiuroPodrozy biuro = new BiuroPodrozy(Nazwa);
            foreach (Podroze l in loty)
            {
                Podroze lot = (Podroze)l.Clone();
                biuro.DodajLot(lot);
            }
            return biuro;
        }

        /// <summary>
        /// Metoda wypisująca nazwę, oraz loty z biura podróży
        /// </summary>
        /// <returns>Wypisana nazwa oraz lista podróży</returns>
        public override string ToString()
        {
            StringBuilder sb = new StringBuilder();
            sb.AppendLine($"Nazwa: {Nazwa}");
            foreach(Podroze lot in loty)
            {
                sb.AppendLine(lot.ToString());
            }

            return sb.ToString();
        }
    }
}

File: Travel managment system - JAVA/src/com/company/BiuroPodrozy.java
package com.company;

import java.io.*;
import java.sql.*;
import java.text.NumberFormat;
import java.util.ArrayList;
import java.util.Objects;

/**
 * Klasa BiuroPodróży przechowująca podróże, korzystająca z interfejsu IBiuro oraz ICloneable.
 */
public class BiuroPodrozy implements IBiuro,Cloneable, Serializable {
    /**
     * Zmienna nazywająca biuro podróży.
     */
    private String nazwa;
    /**
     * Zmodyfikowane pole loty, które jest listą.
     */
    private ArrayList<Podroz> loty;

    /**
     * Konstruktor nieparametryczny tworzący nową listę 'loty' i ustawiający pole 'nazwa'.
     * @param nazwa  nadająca nazwę biura.
     */
    public BiuroPodrozy(String nazwa) {
        this.nazwa = nazwa;
        loty = new ArrayList<>();
    }

    public String getNazwa() {
        return nazwa;
    }

    public ArrayList<Podroz> getLoty() {
        return loty;
    }

  /**
     * Metoda dodająca wybraną podróż.
     * @param t wybrany lot do dodania.
     */
    @Override
    public void DodajLot(Podroz t) {
        loty.add(t);
    }

    /**
     * Metoda usuwająca wybraną podróż.
     * @param t wybrany lot do usunięcia.
     */
    @Override
    public void UsunLot(Podroz t) {
        loty.remove(t);
    }

    /**
     * Metoda usuwająca podróże z wybranym peselem.
     * @param PESEL wybrany pesel.
     */
    @Override
    public void UsunLot(String PESEL) {
        for (Podroz l:loty) {
            if (Objects.equals(l.pesel, PESEL))
            {
                loty.remove(l);
                break;
            }
        }
    }

    /**
     * Metoda sprawdzająca czy wybrana podróż znajduje się w biuerze podróży.
     * @param podroz wybrana podróż.
     * @return prawda lub fałsz.
     */
    @Override
    public boolean CzyPodrozJestWBiurzePodrozy(Podroz podroz) {
        for (Podroz l:loty) {
            if (l.equals(podroz))
                return true;
        }
        return false;
    }

    /**
     * Metoda, która wyszukuje podróż przez podanie nazwy lotniska z którego jest wylot.
     * @param wylot wybrane lotnisko.
     * @return lista podróży z wybranego lotniska.
     */
    @Override
    public ArrayList<Podroz> WyszukajPodroze(String wylot) {
        ArrayList<Podroz> newList = new ArrayList<>();
        for (Podroz l:loty) {
            if (l._cel.toString().contains(wylot))
            {
                newList.add(l);
            }
        }
        return newList;
    }

    /**
     * Metoda, która sortuje loty (po sygnaturze).
     */
    public void Sortuj() {
        loty.sort(Podroz::compareTo);
    }

    /**
     * Metoda, która sortuje loty (po peselu).
     */
    public void SortujPoPESEL() {
        loty.sort(new PESELComparator());
    }

    /**
     * Metoda klonująca biuro podróży. W razie niepowodzenia wyrzuca wyjątek.
     * @return klon podróży.
     */
    @Override
    public BiuroPodrozy clone() {
        try {
            return (BiuroPodrozy) super.clone();
        } catch (CloneNotSupportedException e) {
            throw new AssertionError();
        }
    }

    public void a(BiuroPodrozy b)
    {
        this.nazwa = b.nazwa;
        this.loty = b.loty;
    }

    /**
     * Metoda zapisująca biuro podróży do pliku.
     * @param nazwa nazwa biura podróży.
     */
    public void zapisz(String nazwa){
        try {
            FileOutputStream fileOutputStream = new FileOutputStream(nazwa);
            ObjectOutputStream objOutputStream = new ObjectOutputStream(fileOutputStream);

            objOutputStream.writeObject(this);
            fileOutputStream.close();
            objOutputStream.close();

        } catch ( IOException e ) {
            e.printStackTrace();
        }
    }

    /**
     * Metoda odczytująca biuro podróży z pliku.
     * @param nazwa nazwa biura.
     * @return biuro podrózy.
     */
    public static BiuroPodrozy odczytaj(String nazwa){
        BiuroPodrozy biuro = null;

        try {
            FileInputStream fileInputStream = new FileInputStream(nazwa);
            var objectInputStream = new ObjectInputStream(fileInputStream);

            biuro = (BiuroPodrozy) objectInputStream.readObject();

            objectInputStream.close();
            fileInputStream.close();

        } catch (IOException | ClassNotFoundException e) {
            e.printStackTrace();
        }
        return biuro;
    }

    /**
     * Medota zapisująca wybrane biuro podróży do bazy danych.
     * @param biuro wybrane biuro podróży.
     * @throws SQLException w razie niepowodzenia wyrzuca SQLEException.
     */
    public void ZapiszDoBazy(BiuroPodrozy biuro) throws SQLException {
        Connection connection = DriverManager.getConnection("jdbc:mysql://localhost:3306/biuro-podrozy","root","mypass");
        Statement statement = connection.createStatement();
        biuro.Sortuj();
        for(Podroz l:loty){
            String sqlPodroz = "INSERT INTO podroz (Sygnatura,PESEL, SpecjalnePotrzeby, Ubezpieczenie, DodatkowyBagaz, PrzelecianeKilometry, IloscDzieci, IloscDoroslych, Klasa, Bilet, Cel, Koszt) VALUES  ('"+l.sygnatura+"', '"+l.pesel+"', '"+l.specjalnePotrzeby+"', '"+l.ubezpieczenie+"', '"+l.dodatkowyBagaz+"', '"+l.przelecianeKilometry+"', '"+l.iloscDzieci+"', '"+l.iloscDoroslych+"', '"+l._klasa+"', '"+l._bilet+"', '"+l._cel+"', '"+ NumberFormat.getCurrencyInstance().format(l.PoliczKoszt())+"')";
            String sqlOsoba = "INSERT INTO osoba (Imie, Nazwisko, Adres, KodPocztowy, Telefon, Email, PESEL) VALUES  ('"+l.imie+"', '"+l.nazwisko+"', '"+l.adres+"', '"+l.kodPocztowy+"', '"+l.telefon+"', '"+l.email+"', '"+l.pesel+"')";
            statement.executeUpdate(sqlOsoba);
            statement.executeUpdate(sqlPodroz);
        }
        System.out.println("Zapisano do bazy :)");
    }

    /**
     * Medota odczytująca biuro podróży z bazy danych. W razie niepowodzenia lub niepoprawnych danych wyrzuca wyjątki.
     * @return biuro.
     * @throws SQLException wyjątek.
     * @throws ZlyTelefonException wyjątek.
     * @throws ZlyKodPocztowyException wyjątek.
     * @throws ZlyPeselException wyjątek..
     */
    public static com.company.BiuroPodrozy OdczytajZBazy() throws SQLException, ZlyTelefonException, ZlyKodPocztowyException, ZlyPeselException {
        Connection connection = DriverManager.getConnection("jdbc:mysql://localhost:3306/biuro-podrozy","root","mypass");
        Statement statement = connection.createStatement();
        ResultSet resultSet = statement.executeQuery(("SELECT osoba.Imie, osoba.Nazwisko, osoba.Adres, osoba.KodPocztowy, osoba.Telefon, osoba.Email, osoba.PESEL, podroz.SpecjalnePotrzeby, podroz.Ubezpieczenie, podroz.DodatkowyBagaz, podroz.PrzelecianeKilometry, podroz.IloscDzieci, podroz.IloscDoroslych, podroz.Klasa, podroz.Bilet, podroz.Cel FROM podroz INNER JOIN osoba ON podroz.PESEL=osoba.PESEL;"));

        com.company.BiuroPodrozy noweBiuro=new com.company.BiuroPodrozy("OdczytaneBiuro");
        while(resultSet.next()){
            Podroz p1 = new Podroz(resultSet.getString("Imie"),
                    resultSet.getString("Nazwisko"),
                    resultSet.getString("Adres"),
                    resultSet.getString("KodPocztowy"),
                    resultSet.getString("Telefon"),
                    resultSet.getString("Email"),
                    resultSet.getString("PESEL"),
                    Boolean.parseBoolean(resultSet.getString("SpecjalnePotrzeby")),
                    Boolean.parseBoolean(resultSet.getString("Ubezpieczenie")),
                    Boolean.parseBoolean(resultSet.getString("DodatkowyBagaz")),
                    Boolean.parseBoolean(resultSet.getString("PrzelecianeKilometry")),
                    Integer.parseInt(resultSet.getString("IloscDzieci")),
                    Integer.parseInt(resultSet.getString("IloscDoroslych")),
                    Podroz.klasa.valueOf(resultSet.getString("Klasa")) ,
                    Podroz.bilet.valueOf(resultSet.getString("Bilet")),
                    Podroz.cel.valueOf(resultSet.getString("Cel")));
            noweBiuro.DodajLot(p1);
        }
        return noweBiuro;
    }

    /**
     * Metoda wypisująca nazwę, oraz loty z biura podróży.
     * @return nazwa biura oraz loty.
     */
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("Nazwa: ").append(nazwa).append("\n");
        for(Podroz lot: loty)
        {
            sb.append(lot.toString()).append("\n");
        }
        return sb.toString();
    }
}
